lib/IDE/CodeCompletionResultPrinter.cpp

```diff
@@ -135,29 +157,30 @@ class AnnotatingResultPrinter {
     }
   }
 
-  void printCallArg(ArrayRef<CodeCompletionString::Chunk> chunks) {
-    OS << "<callarg>";
-    for (auto i = chunks.begin(), e = chunks.end(); i != e; ++i) {
+  void printNestedGroup(StringRef tag,
+                        ArrayRef<CodeCompletionString::Chunk>::iterator &i,
+                        ArrayRef<CodeCompletionString::Chunk>::iterator e) {
+    OS << "<" << tag << ">";
+    auto nestingLevel = i->getNestingLevel();
+    // Skip the "Begin" chunk.
+    assert(i != e && !i->hasText() &&
+           CodeCompletionString::Chunk::chunkStartsNestedGroup(i->getKind()));
+    ++i;
+    for (;i != e && !i->endsPreviousNestedGroup(nestingLevel); ++i) {
       using ChunkKind = CodeCompletionString::Chunk::ChunkKind;
-
       if (i->is(ChunkKind::CallArgumentTypeBegin)) {
-        OS << "<callarg.type>";
-        auto nestingLevel = i->getNestingLevel();
-        ++i;
-        for (; i != e; ++i) {
-          if (i->endsPreviousNestedGroup(nestingLevel))
-            break;
-          if (i->hasText())
-            printTextChunk(*i);
-        }
-        OS << "</callarg.type>";
-        if (i == e)
-          break;
+        printNestedGroup("callarg.type", i, e);
+        continue;
+      }
+      if (i->is(ChunkKind::ParameterDeclTypeBegin)) {
+        printNestedGroup("param.type", i, e);
+        continue;
       }
-
       printTextChunk(*i);
     }
-    OS << "</callarg>";
+    OS << "</" << tag << ">";
+    // Move the cursor to the last chunk printed.
+    --i;
```

Comment:
We should add a doc comment that explains how `printNestedGroup` modifies `i`.  The behaviour is surprising if you don't realize it's so that you can immediately call "++i" for the next loop iteration.  Most algorithms that consume an iterator would not do this.

Reply:
Actually, I really didn't like this. Changed it to "Advance `i` to just the past the group" and the call site like
```cpp
auto i = chunks.begin(), e = chunks.end();
while (i != e) {
  ...
}
```